### ISSUE TITLE
[@lit-labs/react]: awkward types when component event name clashes with built-in event name

### ISSUE THREAD

**Issue by WickyNilliams** [lit/lit#2592]
# Description

If you create a react wrapper with a prop name like `onInput` or `onChange` then the resulting types force you to accept a union as the type for your param.

# Steps to Reproduce

1. Write this code

```ts
import { render } from "react-dom";
import React from "react";
import { createComponent } from "@lit-labs/react";
import { MyInput } from "./web-component";

const MyReactInput = createComponent(React, "my-input", MyInput, {
  onInput: "input"
});

function handleInput(e: Event) {
  console.log(e);
}

render(
  <MyReactInput label="test" onInput={handleInput} />, // error here!
  document.getElementById("root")
);

```

2. See this output...

## Live Reproduction Link

Sorry, i went with codesandbox rather than stackblitz since i already had a sandbox i could fork for convenience 

https://codesandbox.io/s/cool-liskov-7sv9eo?file=/src/index.tsx

## Expected Results

I should not have to use a union to type my event handlers, `Event` should suffice (or whatever is idiomatic in React/TS world)

## Actual Results

```
Type '(e: Event) => void' is not assignable to type '((e: Event) => unknown) & FormEventHandler<HTMLElement>'.
```

Can be worked around by typing the event handler as:

```ts
function handleInput(e: Event | FormEvent) {
  console.log(e);
}
```

## Browsers Affected

N/A, typescript issue

- [ ] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Safari 11
- [ ] Safari 10
- [ ] IE 11

**Comment by WickyNilliams** [comment]
I wonder whether you could do something like `Omit` from `HTMLAttributes` all the given event prop names like

```ts
Omit<React.HTMLAttributes<HTMLElement>, keyof Events<E>>
```

on this line https://github.com/lit/lit/blob/b5f1b9ed432ae1961fd182fc71bdfa9f3208624f/packages/labs/react/src/create-component.ts#L136

**Comment by taylor-vann** [comment]
@WickyNilliams thanks for submitting!

Couple of clarifying questions / statements

I think what's happening here is you nailed the most confusing property in the react ecosystem.

Here's a modified example of your submission using `doInput` rather than `onInput` and the typing works :)
https://codesandbox.io/s/dreamy-chatelet-3wzxd3?file=/src/index.tsx

In short:
React `onInput` !== DOM `input`, very far from it

it's a specifically reserved event in the React DOM Event Plugin. Same thing for `onChange` and a few other React Ecosystem specific functions. There will always require a union type required to work with other event types.
 
Changing the event name to properties not reserved by react should fix this :)
 
 I like your suggestion:
 ```
 Omit<React.HTMLAttributes<HTMLElement>, keyof Events<E>>
 ```

I'm working on a PR to improve react support in the labs package. These kinds of reserved properties should flag. Might end up adding it to our list of omitted properties as well.
 
 If the name change does not satisfy your challenge let me know and we'll find another solution

**Comment by WickyNilliams** [comment]
Yeah I get the cause, I've worked with react for a long time :) 

Unfortunately I don't think changing the event name is viable. I want the react experience to be as close to idiomatic as possible for devs consuming my design system. So I want to stick with prop names that _feel_ like a react component - if I were writing a react component with no underlying WC , I'd use names like `onChange` and `onInput`. 

Omitting the event name seems workable to me? I can't imagine a scenario where you'd want react's types for `onChange`/`onInput`/`onWhatever` if you're explicitly declaring you have your own event of the same name. Or am I missing something?

**Comment by taylor-vann** [comment]
I don't think there's anything wrong with that intention at all :) But I feel like there are conflicts in the approach (all semantics though, apologies ahead of time)

> I want the react experience to be as close to idiomatic as possible for devs consuming my design system. So I want to stick with prop names that feel like a react component 

I feel like if that is the intention `FormEventHandler<HTMLElement>` would be the correct type for all `onInput` and `onChange` events in React. Seeing an `Event` would be the odd support case. We're in React land and that what react does :)

The challenge is this component is a React Component that returns a ref to a webcomponent. So my logic would be:
- webcomponent properties are mapped onto react properties
- consider a webcomponent's `input` is not the same as a react's `onInput`
- provide `input` as a property in react-land because that differs it from the `onInput` property

https://codesandbox.io/s/late-cdn-35xl5w?file=/src/index.tsx

I feel the other approach imiplies that the wrapper should support / translate an `onInput` react events to `input` webcomponent events. However that that would require something beyond typecasting, they're very different event types for different kinds of messages

**Comment by WickyNilliams** [comment]
I understand the point you're making, but I don't fully agree. 

I think there is a thin line between being helpful with types and being obstructive. Ultimately the props I _explicitly define_ should take precedence over anything else, and I shouldnt have to compromise my API choices. If I say "this is my prop name", it should not assume anything about that prop. 

> consider a webcomponent's input is not the same as a react's onInput 

This is my thinking, they are not the same so I should not be forced to use the same type

If I create a react component (with no underlying web component) with an `onChange` prop, it does not inherit any type from `HTMLElement`, it uses my own. If I were to define a `focus()` method on my web component (outside of react), I am not forced to use the native html type for that method, my type is the one that is exposed. This feels the same. 

I think this is a case of trying to be too clever with types instead of doing what I, the developer, want. I should be in control. It reduces the usefulness of this library if it makes choices for me that I cannot override.

**Comment by taylor-vann** [comment]
Kewl. I understand the frustration and the sentiment. This swings back around to my previous statment "you nailed the most confusing property in the react ecosystem."

Couple points to address:
-  I agree with you and think that what a user defines should take precedence. Your prop types should 'win'. And they do, initially. They will get past the `props` as the argument of a function component.
- I believe you're confusing the onInput event with the onChange event in the React ecosystem. A 'react-like' api would use `onChange` in conjunction with the HTMLInputElement's `oninput`, ie: `{onChange: "input"}`
- Lastly and most important: It's not the webcomponent or this wrapper that is throwing this error or preventing you from using `onInput`. This is caused by the pragma after the JSX / TSX conversion.

```
 `<input onChange{input}>`
 
 to ...
 
 React.createComponent('input', {onChange: input});
 ^^^^^ this flags the 'onChange' parameter
```

As demonstrated in the example below, regular react components flag an `EventListener` as an invalid property when used for `onInput`, `onSubmit`, or `onChange`. However `input` still allows an EventListener.

https://codesandbox.io/s/affectionate-sea-1fl5nj?file

React itself is preventing this. So when your webcomponent wrapper flags `onInput` as invalid, that is React working as intended.

> I think this is a case of trying to be too clever with types instead of doing what I, the developer, want. I should be in control.

I agree with this :) But as I've hopefully demonstrated, this behavior is not up to lit or our react wrapper :( The `onChange` and `onSubmit` debate has been a source of spicy takes for years.

>  It reduces the usefulness of this library if it makes choices for me that I cannot override.

I disagree. This is not a reflection of the quality or caliber of webcomponents or this wrapper. We are all subject to the react silo when we use react, the dom will always be a rendering target not a source of truth, typing or otherwise. I believe this sentiment should be directed towards the React library.

**Comment by WickyNilliams** [comment]
Perhaps I'm not explaining things well because I'm typing on mobile. I'll get back to this Monday, I don't want to take up too of your saturday! Have a good weekend.

**Comment by taylor-vann** [comment]
Sounds good! Talk Monday :)

**Comment by WickyNilliams** [comment]
Wonderful, thank you! 

I had been meaning to get back to this, but I've been entirely focused on other tasks the last few weeks ahead of a big release. Sadly this fell to the wayside. 

If you need me to do some testing of your PR with my design system, let me know and I'll give it a whirl.

Thanks again!

**Comment by taylor-vann** [comment]
I'll tag you in the PR or something :) I used the `Omit` bit you suggested!

**Comment by taylor-vann** [comment]
resolved in #2648
